### PR TITLE
perf: improve the API response time

### LINKED ISSUE #87: perf: improve the API response time
## Summary

We are improving the evaluation process on the server to evaluate only the flags that have been changed.
This will improve the load on the server and the response time since it will only return the flags that have been changed.

Currently, when a flag changes on the server, it evaluates all flags again and sends the latest evaluations to the client SDK.
The SDK checks the response and does nothing if the response is empty. Otherwise, it deletes all the data in the local storage and saves the latest evaluations.

### When the SDK is initialized

1- Change the `featureTag` setting to be optional in the `BKTConfig`
2- Save the `featureTag` in the `localStorage` if it is configured in the `BKTConfig`
3- Clear the `userEvaluationsID` in the `localStorage` if the `featureTag` changes

### When sending requests

The following new fields will be added to `get_evaluations` request.

- **evaluatedAt**: the last time the user was evaluated. The server will return in the get_evaluations response (`UserEvaluations.CreatedAt`), and it must be saved in the client
- **userAttributesUpdated**: when the user attributes change via the `customAttributes` interface, the userAttributesUpdated field must be set to true in the next request

### When getting the response (Update conditions)

The server will return the new following fields in the response.

- **forceUpdate**: a boolean that tells the SDK to delete all the current data and save the latest evaluations from the response
- **archivedFeatureIds**: a list of feature flag ids that were archived on the admin console. All those feature flags must be deleted from the local storage.

#### forceUpdate is True

1- Delete all the evaluations from local storage, and save the latest evaluations from the response into the local storage
2- Save the `UserEvaluations.CreatedAt` in the response as `evaluatedAt` in the `localStorage`

#### forceUpdate is false

1- Check the evaluation list  in the response and **upsert** them in the `localStorage` if the list is not empty
2- Check the `archivedFeatureIds` list and delete them from the `localStorage` if is not empty
3- Save the `UserEvaluations.CreatedAt` in the response as `evaluatedAt` in the `localStorage`

### LINKED ISSUE #93: chore: add the sdkVersion in the requests
To help monitor the APIs by `sdkVersion`, we need to add it to the `GetEvaluations` and `RegisterEvents` APIs.

### Things to do

- Add the `sdkVersion` in the API request.

#### GetEvaluations

https://github.com/bucketeer-io/javascript-client-sdk/blob/6d600dd4998e2483d0c24aab28cc23bbddb9f33e/src/internal/model/request/GetEvaluationsRequest.ts#L4-L9

#### RegisterEvents

https://github.com/bucketeer-io/javascript-client-sdk/blob/6d600dd4998e2483d0c24aab28cc23bbddb9f33e/src/internal/model/request/RegisterEventsRequest.ts#L3-L5

### PR DESCRIPTION
**Pull request by yshrsmz**
resolve #87 
resolve #93 

## todo
- [x] change createdAt and evaluatedAt to string
- [x] add e2e test cases if possible

### CODE REVIEW

File: src/internal/evaluation/EvaluationStorage.ts [hunk 1]
```diff
@@ -5,13 +5,39 @@ export interface EvaluationEntity {
   userId: string
   currentEvaluationsId: string | null
   evaluations: Record<string /* featureId */, Evaluation>
+  currentFeatureTag: string | null
+  evaluatedAt: number | null
+  userAttributesUpdated: boolean
```
Comment (yshrsmz) [inline src/internal/evaluation/EvaluationStorage.ts:10]: @cre8ivejp Do you want to persist `userAttributesUpdated`?

I mean, when the SDK fails to notify the backend that `userAttributesUpdated: true` and a user close the app, do you want to send a request with `userAttributesUpdated: true` upon the next launch?

File: src/internal/model/request/GetEvaluationsRequest.ts [hunk 2]
```diff
@@ -1,9 +1,17 @@
...
+  userAttributesUpdated: boolean
+}
+
 export interface GetEvaluationsRequest {
   tag: string
   user: User
   userEvaluationsId: string
   sourceId: SourceID
+  userEvaluationCondition: UserEvaluationCondition
+  sdkVersion: string
```
Comment (cre8ivejp) [inline src/internal/model/request/GetEvaluationsRequest.ts:16]: Can you also add the `sdkVersion` in the `RegisterEventsRequest`?

https://github.com/bucketeer-io/javascript-client-sdk/blob/c98ee30f06ece69f223a9bf25d3e6de55d4a8b03/src/internal/model/request/RegisterEventsRequest.ts#L3C18-L4

Comment (yshrsmz) [reply src/internal/model/request/GetEvaluationsRequest.ts:16]: done: [`1bd7cdf` (#91)](https://github.com/bucketeer-io/javascript-client-sdk/pull/91/commits/1bd7cdfa298e08463a486fdd6974b4fc1eb514cb)

File: e2e/BKTClient.spec.ts [hunk 3]
```diff
@@ -60,4 +64,90 @@ suite('e2e/BKTClientTest', () => {
...
+
+      const current = evaluationStorage.storage.get()
+
+      assert(current != null)
+
+      // update evaluations manually, and check if evaluation1 is deleted after fetchEvaluations
+      evaluationStorage.storage.set({
+        ...current,
+        evaluations: { [evaluation1.featureId]: evaluation1 },
+        evaluatedAt: '1',
```
Comment (yshrsmz) [inline e2e/BKTClient.spec.ts:86]: It seems like just changing `evaluatedAt` to a super-old value does not make force-update.

this `/get_evaluations` request

```json
{
  "user":{"id":"bucketeer-js-user-id-1","attributes":{}},
  "userEvaluationsId":"3227641913513702639",
  "userEvaluationCondition":{"evaluatedAt":"1","userAttributesUpdated":false},
  "tag":"javascript",
  "sourceId":7,
  "sdkVersion":"2.0.2"
}
```

results in the response below

```json
{
    "state": "FULL",
    "evaluations": {
        "id": "no_evaluations",
        "evaluations": [],
        "createdAt": "1689779952",
        "archivedFeatureIds": [],
        "forceUpdate": false
    },
    "userEvaluationsId": "3227641913513702639"
}
```

File: e2e/BKTClient.spec.ts [hunk 4]
```diff
@@ -60,4 +64,90 @@ suite('e2e/BKTClientTest', () => {
...
+
+      const current = evaluationStorage.storage.get()
+
+      assert(current != null)
+
+      // update evaluations manually, and check if evaluation1 is deleted after fetchEvaluations
+      evaluationStorage.storage.set({
+        ...current,
+        evaluations: { [evaluation1.featureId]: evaluation1 },
+        currentEvaluationsId: null,
```
Comment (yshrsmz) [inline e2e/BKTClient.spec.ts:126]: On the other hand, request with empty evaluationsId  works as expected.

request:

```json
{
  "user":{"id":"bucketeer-js-user-id-1","attributes":{}},
  "userEvaluationsId":"",
  "userEvaluationCondition":{"evaluatedAt":"1689779952","userAttributesUpdated":false},
  "tag":"javascript",
  "sourceId":7,
  "sdkVersion":"2.0.2"
}
```

response:

```json
{
    "state": "FULL",
    "evaluations": {
        "id": "11556537758303297585",
        "evaluations": [
            {
                "id": "feature-javascript-e2e-boolean:2:bucketeer-js-user-id-1",
                "featureId": "feature-javascript-e2e-boolean",
                "featureVersion": 2,
                "userId": "bucketeer-js-user-id-1",
                "variationId": "3fd13e1e-a639-4a36-bccb-69a23445ea48",
                "variation": {
                    "id": "3fd13e1e-a639-4a36-bccb-69a23445ea48",
                    "value": "false",
                    "name": "",
                    "description": ""
                },
                "reason": {
                    "type": "OFF_VARIATION",
                    "ruleId": ""
                },
                "variationValue": "false",
                "variationName": ""
            },
            {
                "id": "feature-js-e2e-boolean:3:bucketeer-js-user-id-1",
                "featureId": "feature-js-e2e-boolean",
                "featureVersion": 3,
                "userId": "bucketeer-js-user-id-1",
                "variationId": "4fab39c8-bf62-4a78-8a10-1b8bc3dd3806",
                "variation": {
                    "id": "4fab39c8-bf62-4a78-8a10-1b8bc3dd3806",
                    "value": "true",
                    "name": "",
                    "description": ""
                },
                "reason": {
                    "type": "DEFAULT",
                    "ruleId": ""
                },
                "variationValue": "true",
                "variationName": ""
            },
            {
                "id": "feature-js-e2e-double:3:bucketeer-js-user-id-1",
                "featureId": "feature-js-e2e-double",
                "featureVersion": 3,
                "userId": "bucketeer-js-user-id-1",
                "variationId": "2d4a213c-1721-434b-8484-1b72826ece98",
                "variation": {
                    "id": "2d4a213c-1721-434b-8484-1b72826ece98",
                    "value": "2.1",
                    "name": "",
                    "description": ""
                },
                "reason": {
                    "type": "DEFAULT",
                    "ruleId": ""
                },
                "variationValue": "2.1",
                "variationName": ""
            },
            {
                "id": "feature-js-e2e-int:3:bucketeer-js-user-id-1",
                "featureId": "feature-js-e2e-int",
                "featureVersion": 3,
                "userId": "bucketeer-js-user-id-1",
                "variationId": "6079c503-c281-4561-b870-c2c59a75e6a6",
                "variation": {
                    "id": "6079c503-c281-4561-b870-c2c59a75e6a6",
                    "value": "10",
                    "name": "",
                    "description": ""
                },
                "reason": {
                    "type": "DEFAULT",
                    "ruleId": ""
                },
                "variationValue": "10",
                "variationName": ""
            },
            {
                "id": "feature-js-e2e-json:3:bucketeer-js-user-id-1",
                "featureId": "feature-js-e2e-json",
                "featureVersion": 3,
                "userId": "bucketeer-js-user-id-1",
                "variationId": "8b53a27b-2658-4f8c-925e-fb277808ed30",
                "variation": {
                    "id": "8b53a27b-2658-4f8c-925e-fb277808ed30",
                    "value": "{ \"key\": \"value-1\" }",
                    "name": "",
                    "description": ""
                },
                "reason": {
                    "type": "DEFAULT",
                    "ruleId": ""
                },
                "variationValue": "{ \"key\": \"value-1\" }",
                "variationName": ""
            },
            {
                "id": "feature-js-e2e-string:5:bucketeer-js-user-id-1",
                "featureId": "feature-js-e2e-string",
                "featureVersion": 5,
                "userId": "bucketeer-js-user-id-1",
                "variationId": "87e0a1ef-a0cb-49da-8460-289948f117ba",
                "variation": {
                    "id": "87e0a1ef-a0cb-49da-8460-289948f117ba",
                    "value": "value-1",
                    "name": "",
                    "description": ""
                },
                "reason": {
                    "type": "DEFAULT",
                    "ruleId": ""
                },
                "variationValue": "value-1",
                "variationName": ""
            }
        ],
        "createdAt": "1689779952",
        "archivedFeatureIds": [],
        "forceUpdate": true
    },
    "userEvaluationsId": "3227641913513702639"
}
```

**Review by yshrsmz — COMMENTED** [review]
Added 2 e2e test cases, but 1 case is not working as expected.

@cre8ivejp Do you have any idea what's going on? 🙏

File: e2e/BKTClient.spec.ts (hunk 3 above)
Comment (cre8ivejp) [reply e2e/BKTClient.spec.ts:86]: The condition is when the `userEvaluationsId` is different and the `evaluatedAt` is older than 31 days.
So we need to use a random string in the `userEvaluationsId` to make it work.

Sorry for my mistake.

Comment (yshrsmz) [reply e2e/BKTClient.spec.ts:111]: Thanks, fixed: [`4ac8be3` (#91)](https://github.com/bucketeer-io/javascript-client-sdk/pull/91/commits/4ac8be3adce5480619b0cd12a015d76ea0183807)

**Review by cre8ivejp — APPROVED** [review]
Amazing work!
Thank you!

**Comment by yshrsmz** [comment]
Thanks!

File: e2e/BKTClient.spec.ts (hunk 3 above)
Comment (duyhungtnn) [reply e2e/BKTClient.spec.ts:111]: Hi @yshrsmz  , I updated those changes to iOS. 
Is there a way to e2e test the `upsert evaluation` cases? I was looking at your PR. But I did not find it. 

cc: @cre8ivejp

Comment (yshrsmz) [reply e2e/BKTClient.spec.ts:111]: @duyhungtnn It's impossible to configure `archivedFeatureIds` and other stuff for e2e testing. So I just wrote forceUpdate cases.

`upsert` part is tested through unit tests, so it should be fine.

Comment (duyhungtnn) [reply e2e/BKTClient.spec.ts:111]: thank you for confirming.

**Pull request merged**